### PR TITLE
feat(#137): CPython traceback fixture + log parser alignment

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

The log parser's Python traceback handling only recognised `File "..."` lines, missing the interleaved source-code preview lines that real CPython tracebacks include. This broke multi-frame elision (source lines were counted as extra frames) and wasted 3–4× tokens for AI agents reading Python exception output.

## Changes

**Parser (`crates/rskim/src/cmd/log.rs`)**
- Added `is_python_continuation()` — pure predicate that detects source-preview and PEP 657 caret lines; runs before `classify_log_line` to prevent source lines containing log keywords (e.g. `"ERROR: bad input"`) from being misclassified
- `VecDeque<String>` replaces `Vec<String>` for `pending_stack` — O(1) `pop_front` instead of `O(n)` `remove(0)` for cap eviction
- `in_python_frame: bool` flag — set on each `File "..."` match, cleared on any non-continuation line
- Traceback header (`Traceback (most recent call last)`) attached to the preceding log entry (or created as unstructured if no entry exists yet)
- Chained exception separators (`The above exception was the direct cause…`, `During handling of the above exception…`) flush the pending stack and push as unstructured entries
- `found_structured = true` now set inside the `RE_LOG_STACK_TRACE` branch so standalone tracebacks (no log-level prefix) are parsed as structured output

**Fixtures**
- `stack_trace_python.txt` — updated to CPython 3.11+ format with source-preview lines per frame
- `stack_trace_python_chained.txt` — new fixture: chained exception with two `Traceback` blocks separated by `The above exception was the direct cause…`
- `stack_trace_python_pep657.txt` — new fixture: PEP 657 fine-grained error location with caret lines

**Tests — 14 new tests**
- Group A: source-preview attachment, logical frame count not inflated by preview lines
- Group B: chained exception — Traceback headers attached, correct elision count, INFO not contaminated
- Group C: PEP 657 caret lines attached, not counted as frames
- Group D: regression guards — no extra entries from preview lines, `found_structured` gate, misclassification guard, cap behaviour with preview lines, Traceback-only header
- Group F: rendered output coherent (ValueError, elision marker), chained output complete

## Breaking Changes

None — existing Java/Node.js stack trace behaviour unchanged. All 3680 rskim tests pass.

## Reviewer Focus Areas

- State machine control flow in `try_parse_regex_logs` — verify branch ordering (Step 2 RE_LOG_STACK_TRACE → Step 3 continuation → Step 4 strip_timestamp → Step 5 Traceback header → Step 6 separator → Step 7 reset → Step 8 flush+classify)
- `is_python_continuation()` predicate — verify it catches source-preview, PEP 657 carets, and `[Previous line repeated]`-style lines without false positives on non-Python indented content
- Chained exception separator detection — verify explicit flush trigger and that the separator line is pushed as unstructured (no level)
- Fixture content — verify realistic CPython 3.11+ format

### CODE REVIEW

**Comment by dean0x** [comment]
# Code Review: PR #271

## Findings (≥80% Confidence)

### 🛑 Blocking Issues

#### 1. **Cyclomatic Complexity Exceeds Threshold** — `log.rs:398` (92% confidence)
The function has ~22 cyclomatic complexity (2x threshold) with 8 branches and 6 mutable state variables.

**Fix**: Extract the 4-repeating flush-guard pattern into a helper:
```rust
fn try_flush(
    all_entries: &mut Vec<(Option<String>, String)>,
    pending_stack: &mut VecDeque<String>,
    elided: &mut usize,
) {
    if !pending_stack.is_empty() && !all_entries.is_empty() {
        flush_stack_frames(all_entries, pending_stack, elided);
    }
}
```
This cuts ~16 lines and ~8 decision points. Applies **ADR-001** (fix noticed issues immediately).

---

#### 2. **Missing Direct Unit Test for `is_python_continuation`** — `log.rs:360` (85% confidence)
The new helper is tested only indirectly through integration tests. A focused unit test verifies boundary behavior when the regex changes in the future.

**Suggested tests**:
```rust
#[test]
fn test_is_python_continuation_accepts_indented_source() {
    assert!(is_python_continuation("    result = int(value)"));
}

#[test]
fn test_is_python_continuation_rejects_empty() {
    assert!(!is_python_continuation(""));
}

#[test]
fn test_is_python_continuation_rejects_unindented() {
    assert!(!is_python_continuation("ValueError: bad input"));
}
```

---

#### 3. **Unbounded Per-Frame Continuation Growth + Missing Invariant Assert** — `log.rs:442-446` (82-85% confidence)
Two concerns:

**a) Memory safety**: Each continuation line appends to the frame String without a per-frame cap. While `PENDING_STACK_CAP` bounds frame count, this violates the O(PENDING_STACK_CAP) invariant. Fix: Add cap (~4 continuations):
```rust
const MAX_CONTINUATIONS_PER_FRAME: usize = 4;
let continuation_count = last_frame.chars().filter(|&c| c == '\n').count();
if continuation_count < MAX_CONTINUATIONS_PER_FRAME {
    last_frame.push('\n');
    last_frame.push_str(trimmed);
}
```

**b) Missing debug_assert**: Add invariant assertion:
```rust
debug_assert!(!pending_stack.is_empty(), "in_python_frame set without pending frame");
```

---

### ⚠️ Should-Fix Issues (80-85% Confidence)

#### 4. **Redundant Regex in Hot Path** — `log.rs:366` (82% confidence)
The `is_python_continuation` function checks `RE_LOG_STACK_TRACE.is_match()`, but Step 2 already proved this line does NOT match. Redundant on deep tracebacks.

**Fix**: Inline the check in the hot path:
```rust
if in_python_frame && !trimmed.is_empty()
    && line.starts_with(|c: char| c.is_ascii_whitespace())
{
    if let Some(last_frame) = pending_stack.back_mut() {
        last_frame.push('\n');
        last_frame.push_str(trimmed);
    }
    continue;
}
```

---

#### 5. **Behavioral Change: Tier Classification for Stack-Trace-Only Input** — `log.rs:435` (82% confidence)
Setting `found_structured = true` changes edge-case tier from Passthrough → Degraded for stack-trace-only input. Output is identical, but consumers checking `is_passthrough()` see different behavior.

**Action**: Choose one:
- Accept the change (intentional improvement: structure WAS found)
- Preserve old behavior by gating on `!all_entries.is_empty()`

---

### 📋 Lower-Confidence Suggestions (60-79%)

- **State machine coupling** (82%) — Consider extracting Steps 3-6 into a separate handler function (future refactor)
- **Test group label skips Group E** (90%) — Renumber "Group F" to "Group E" for sequential consistency

---

## Summary

| Category | Count | Recommendation |
|----------|-------|-----------------|
| Blocking (HIGH/CRITICAL) | 3 | **Fix all** |
| Should-Fix (MEDIUM) | 2 | **Fix all** |
| Suggestions (lower confidence) | 2 | **Consider in follow-up** |

**Overall**: Strong PR with well-tested new functionality. Recommend addressing the 3 blocking issues (complexity, unit test, memory safety) before merge. Applies **ADR-001** throughout — all noticed issues should be fixed immediately, not deferred.

---
*Claude Code Reviewer* | Generated 2026-06-04

**Comment by dean0x** [comment]
**Review Comment: Missing `Copy` derive on FrameContext**

File: `crates/rskim/src/cmd/log.rs` (line 55)

Missing `Copy` derive for consistency. All fields in this enum are `Copy` types (`usize`). Per the PR description's stated convention ("Debug+Clone+Copy+PartialEq+Eq for data-free"), `FrameContext` should derive `Copy` — it carries only `continuation_count: usize`. Other data-free enums in this PR (`CommitParsePhase`, `PytestSection`) correctly derive `Copy`. 

**Fix:**
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum FrameContext {
```

**Confidence:** 82% (MEDIUM - blocking)

---
*Review finding by Claude Code — Cycle 2 inline comments*

**Comment by dean0x** [comment]
**Review Comment: Non-idiomatic string slicing in `parse_verbose_marker`**

File: `crates/rskim/src/cmd/test/pytest.rs` (line 276)

The codebase consistently uses `strip_prefix`/`strip_suffix` (10+ call sites across cargo.rs, pytest.rs, show.rs), but this is the only instance of manual `[..len - len]` slicing. Since `line.ends_with(suffix)` is already confirmed by the `if` chain above, use `strip_suffix` which is both safer and matches codebase convention.

**Fix:**
```rust
// Before:
Some((line[..line.len() - suffix.len()].to_string(), outcome))

// After:
Some((line.strip_suffix(suffix).unwrap().to_string(), outcome))
```

**Confidence:** 92% (MEDIUM - blocking, consistency violation)

---
*Review finding by Claude Code — Cycle 2 inline comments*

**Comment by dean0x** [comment]
**Review Comment: Missing unit tests for `parse_verbose_marker` helper**

File: `crates/rskim/src/cmd/test/pytest.rs` (line 266, function definition)

Missing unit tests for extracted helper. The function has zero direct unit tests — it is only exercised indirectly through `tier1_parse` integration tests. This is the only location where the subtle behavioral improvement (using `strip_suffix` vs `trim_end_matches`) is exercised, leaving a regression risk.

**Fix:** Add 3-4 unit tests covering the three outcome variants and edge cases:

```rust
#[test]
fn test_parse_verbose_marker_passed() {
    let (name, outcome) = parse_verbose_marker("tests/test_a.py::test_one PASSED").unwrap();
    assert_eq!(name, "tests/test_a.py::test_one");
    assert_eq!(outcome, TestOutcome::Pass);
}

#[test]
fn test_parse_verbose_marker_none_for_plain_line() {
    assert!(parse_verbose_marker("some random output").is_none());
}

#[test]
fn test_parse_verbose_marker_name_containing_keyword() {
    // Ensures old trim_end_matches behavior is not re-introduced
    let (name, _) = parse_verbose_marker("test_PASSED_check PASSED").unwrap();
    assert_eq!(name, "test_PASSED_check");
}
```

**Confidence:** 85% (HIGH - blocking, test coverage)

---
*Review finding by Claude Code — Cycle 2 inline comments*

**Comment by dean0x** [comment]
**Review Comment: Missing unit tests for `CaptureState::flush` method**

File: `crates/rskim/src/cmd/test/cargo.rs` (line 242, method definition)

Missing unit tests for new `flush` method. This method centralizes three previously duplicated flush blocks, but has zero direct unit tests — only exercised through nextest integration tests. The nextest fixture has only one STDOUT capture block, leaving edge cases untested (no matching entry, blank accumulated output).

**Fix:** Add direct unit tests for the three documented behaviors:

```rust
#[test]
fn test_capture_state_flush_idle_is_noop() {
    let mut entries = vec![];
    let result = CaptureState::Idle.flush(&mut entries);
    assert_eq!(result, CaptureState::Idle);
}

#[test]
fn test_capture_state_flush_attaches_detail() {
    let mut entries = vec![TestEntry {
        name: "mytest".to_string(),
        outcome: TestOutcome::Fail,
        detail: None,
    }];
    let cap = CaptureState::Capturing {
        test_name: "mytest".to_string(),
        output: "  assertion failed  ".to_string(),
    };
    let result = cap.flush(&mut entries);
    assert_eq!(result, CaptureState::Idle);
    assert_eq!(entries[0].detail.as_deref(), Some("assertion failed"));
}

#[test]
fn test_capture_state_flush_blank_output_is_noop() {
    let mut entries = vec![TestEntry {
        name: "mytest".to_string(),
        outcome: TestOutcome::Fail,
        detail: None,
    }];
    let cap = CaptureState::Capturing {
        test_name: "mytest".to_string(),
        output: "   \n  ".to_string(),
    };
    cap.flush(&mut entries);
    assert!(entries[0].detail.is_none());
}
```

**Confidence:** 82% (HIGH - blocking, test coverage)

---
*Review finding by Claude Code — Cycle 2 inline comments*

**Comment by dean0x** [comment]
## Cycle 2 Review Summary — PR #271

### Overall Status
**3 reviewers, 4 blocking MEDIUM issues identified** (all ≥80% confidence, flagged as inline comments above)

| Category | CRITICAL | HIGH | MEDIUM |
|----------|----------|------|--------|
| **Blocking Issues** | 0 | 2 | 2 |
| **Should Fix** | — | — | 1 |
| **Pre-existing** | — | — | 0 |

### Issues Requiring Action (Inline Comments Posted)

1. **RUST-3 (82%)**: `FrameContext` missing `Copy` derive — consistency with PR convention and other enums in this PR
2. **CONSISTENCY (92%)**: `parse_verbose_marker` byte slicing vs `strip_suffix` — only instance of non-idiomatic slicing in codebase
3. **TESTING-3 (85%)**: No unit tests for `parse_verbose_marker` helper — extracted function with zero direct test coverage
4. **TESTING-4 (82%)**: No unit tests for `CaptureState::flush` method — centralizes duplicated logic but is untested in isolation

### Lower-Confidence Findings (60-79%, Summary Only)

These are actionable improvements but not blocking — consider for Cycle 3 or a follow-up:

- **RUST-4 (80%)**: Removed nextest capture-reset heuristic changes behavior without test coverage. The old "reset on non-indented lines" heuristic was removed entirely — if intentional, document the assumption. If PASS/FAIL lines can appear inside capture blocks, add a test exercising this scenario.

- **COMPLEXITY (85%)**: Three long functions remain in HIGH range (120-143 lines):
  - `try_parse_regex_logs` (143 lines) — Consider extracting `process_line(...)` helper or coarser-grained `ParsePhase` enum
  - `try_parse_nextest` (121 lines) — Extract a `parse_nextest_result()` helper analogous to `parse_verbose_marker` (would reduce ~40 lines of near-duplicate if-let chains)
  - `tier1_parse` (133 lines) — Extract `detect_section_header()` helper for the three `starts_with("===")` checks

- **COMPLEXITY: Nesting (85%)**: Python continuation block reaches depth 4 (lines 477-494 in log.rs) — flatten with early-continue guard on `continuation_count >= MAX`

- **TESTING: nextest detail coverage (82%)**: `test_tier1_nextest_dedup` verifies dedup but not that `entry.detail` is populated after flush — add `assert!(detail.is_some())`

- **RUST: CaptureState::flush linear scan (65%)**: `for entry in entries.iter_mut()` is O(n) per call — for typical test counts this is fine, but a HashMap lookup would be O(1) if test counts grow large

### What Went Well

- **Strong enum state machine design** across all 4 files — each refactored enum eliminates impossible states. Textbook "make illegal states unrepresentable."
- **Excellent rustdoc** — every enum variant has doc comments explaining its role in the state machine.
- **`CaptureState::flush` consuming method** — clean move semantics, centralizes duplicated drain-and-attach blocks.
- **Comprehensive regression test coverage** — 2 new tests for orphaned-frames behavior (lines 448, 557).
- **Zero clippy warnings** — consistent with project quality standards.

### Next Steps

1. **Apply 4 blocking fixes** (all MEDIUM):
   - Add `Copy` derive to `FrameContext`
   - Replace byte slicing with `strip_suffix` in `parse_verbose_marker`
   - Add 3-4 unit tests for `parse_verbose_marker`
   - Add 3 unit tests for `CaptureState::flush`

2. **Consider for Cycle 3** (if time permits):
   - Extract `parse_nextest_result()` helper to collapse PASS/FAIL/SKIP duplication
   - Flatten nesting depth 4 in Python continuation block

3. **Documentation note**: PR description claims "no behavior changes" but `pending_stack.clear()` and `|| all_entries.is_empty()` guard are genuine behavior fixes for orphaned stack frame handling. Consider clarifying this in the PR description.

---

**Cycle Status**: 4 issues identified → 4 inline comments posted → awaiting fixes → Cycle 3 verification

*Claude Code — Cycle 2 automated review*

File: crates/rskim/src/cmd/log.rs [hunk 1]
```diff
@@ -350,19 +377,37 @@ fn extract_json_message(obj: &Value) -> Option<String> {
...
-/// 2. Keep only the **last** 3 frames (preserving their relative order):
-///    `pending_stack.iter().rev().take(3).rev()`.
+/// 2. Keep only the **last** 3 frames (preserving their relative order).
 /// 3. The frames are appended as a newline-joined suffix to the previous
 ///    entry's message.
 /// 4. Elided count: `total_frames.saturating_sub(3)` — accumulated across
 ///    all entries and surfaced in `LogResult::stack_frames_elided`.
+///
+/// # CPython traceback extensions (issue #137)
+/// Each entry in `pending_stack` is a **logical frame** — a `File "..."` line
```
Comment (dean0x) [inline crates/rskim/src/cmd/log.rs:415]: The doc comment's third bullet claims the function "Does NOT itself match the `RE_LOG_STACK_TRACE` pattern", but the actual function body only checks leading whitespace and non-empty trim—it doesn't reference `RE_LOG_STACK_TRACE` at all.

The call-site ordering (Step 2 before Step 3) ensures correctness, but the docstring is misleading. A future contributor reading the docstring alone could make incorrect assumptions about what the function validates.

**Suggested fix:** Clarify that RE_LOG_STACK_TRACE exclusion is enforced by call-site ordering, not by the function itself.

---
Claude Code Review | RUST-MEDIUM (82% confidence)

File: crates/rskim/src/cmd/test/cargo.rs [hunk 2]
```diff
@@ -401,6 +362,28 @@ fn try_parse_nextest(stdout: &str) -> Option<TestResult> {
...
+/// Called for PASS, FAIL, and SKIP lines; the `outcome` discriminates the
+/// variant. Duplicate `(name, outcome)` pairs are silently dropped via `seen`.
+fn push_nextest_entry(
+    rest: &str,
+    outcome: TestOutcome,
+    seen: &mut HashSet<(String, String)>,
+    entries: &mut Vec<TestEntry>,
+) {
+    if let Some(name) = extract_nextest_name(rest) {
+        let key = (name.clone(), format!("{outcome:?}"));
```
Comment (dean0x) [inline crates/rskim/src/cmd/test/cargo.rs]: This line allocates a new String on every nextest entry via `format!("{outcome:?}")`. Since `TestOutcome` has only 3 variants, the Debug representation is fixed ("Pass", "Fail", "Skip"), yet a fresh heap String is allocated for every test entry.

For large test suites (hundreds to thousands of tests), this is hundreds of unnecessary allocations. Since `TestOutcome` already derives `Eq` + `Hash`, use a direct tuple key instead:

```rust
let key = (name.clone(), outcome);
if seen.insert(key) { ... }
```

This eliminates the format allocation and makes dedup semantics explicit.

---
Claude Code Review | PERFORMANCE-MEDIUM (82% confidence)
